### ISSUE TITLE
Link time issue/regression

### ISSUE THREAD

**Issue by arnemertz** [catchorg/Catch2#1247]
## Description
Consider the project at https://github.com/arnemertz/commandline-videostore-cpp which uses a single unit test that compares two strings (actual vs. expected output of small program)

I found out that using Catch for a small project like this has an impact on link times that got severe in the last versions.

Using GCC 7.2.0 I did some measurements and found the following compile and link time differences (Debug mode):

- A version of the project without Catch (branch E0) takes about 3-4 seconds to compile and link.
- A version using Catch 1.8.2 (branch old-catch) takes about 28 seconds, of which 16 seconds can be attributed to the link time. Compilation of the default Catch main function amounts to about 6 seconds.
- A version using Catch 2.2.2 (branch new-catch) takes **145-160 seconds** to build, of which 110-120 seconds are link time and about 20-30 are compilation time for the default Catch main function.

## Versions/Platform
* Catch version: **v2.2.2** (v1.8.2 for comparison) header only.
* Operating System: **Windows + MSYS**
* Compiler+version: **gcc 7.2.0**

**Comment by horenmar** [comment]
You are not the first person to have a massive link-time regression when using some variation of Windows-native gcc, see e.g. #1205. As far as we can tell, this does not affect MSVC nor GCC on Linux, so it is something specific about interplay between GCC and Windows.


Anyway, I don't have MSYS here, but I do have Nuwen MinGW and WSL. Testing these, I get these numbers:

* WSL + g++ 5.4 needs ~5 seconds to compile Catch 1.8.2 main
* WSL + g++ 5.4 needs ~7.6 seconds to compile Catch2 2.2.2 main
* WSL linking against single test file requires ~0.33 s for Catch 1.8.2
* WSL linking against single test file requires ~0.55 s for Catch2 2.2.2

* MinGW g++ 7.3.0 needs 3.8 s to compile Catch 1.8.2 main
* MinGW g++ 7.3.0 needs 8.2 s to compile Catch2 2.2.2 main
* MinGW linking needs 2.3 s to link Catch 1.8.2
* MinGW linking needs 9 s to link Catch2 2.2.2

In other words, MinGW links terribly slowly, but should not link _that_ slowly. Are you using SSD or spinning rust?

**Comment by arnemertz** [comment]
I'm using an SSD. I'll measure on my home PC as soon as I'm out of my current travel-fest.

**Comment by nikodemus** [comment]
Being new to catch I'm not sure if this is a regression, but I submit that

```
$ \time g++ test.cpp -O0 -std=c++11 -o ./test
6.40user 1.73system 0:08.15elapsed 99%CPU (0avgtext+0avgdata 456192maxresident)k
```

for this:

```
#define CATCH_CONFIG_MAIN
#include "catch.hpp"

TEST_CASE("Numbers work")
{
  int a = 4;
  int b = 3;
  int c = a * b;
  CHECK(a < c);
  REQUIRE(c == 12);
}
```

in not reasonable.

```
$ g++ --version
g++ (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
```

The disk is  an SSD and the machine is on i7-7700HQ.

This was on the Windows Linux Subsystem, but g++ is generally working fine and fast here, so I don't think that's a factor.

Skipping link is just as slow:

```
$ \time g++ test.cpp -O0 -std=c++11 -c -o ./test.o
6.00user 1.21system 0:07.24elapsed 99%CPU (0avgtext+0avgdata 456196maxresident)k
0inputs+0outputs (0major+156574minor)pagefaults 0swaps
```

Things become fast if I leave out the #define CATCH_CONFIG_MAIN

**Comment by chris-stones** [comment]
Similar issue here....
This takes 20 seconds to compile, 1 minute, 39 seconds to link.
This is on a 5th generation i7, 8gigs of RAM, SSD.
Compiler used is mingw-w64  x86_64-8.1.0-win32-seh-rt_v6-rev0 (installed from CLion IDE)
Windows-10.
Using GIT version at a575536abe20b58c48863600f8a71e93b4052b81
...Also.. I'm new to this tool.. but i wasnt expecting the output - "All tests passed (1 assertion in 1 test case)".
```
set(PROJ_NAME UnitTests)
project(${PROJ_NAME})
set(CMAKE_CXX_STANDARD 17)
add_executable(${PROJ_NAME} main.cpp)
```
```
#define CATCH_CONFIG_MAIN
#include "catch2/catch.hpp"

TEST_CASE("Trivial test", "[trivial]") {
	REQUIRE(12!=13);
}
```

**Comment by tower120** [comment]
I can confirm - problem in MinGW [tried : x86_64-8.1.0-posix-seh-rt_v6-rev0, x86_64-7.3.0-posix-seh-rt_v5-rev0]. 
The most basic test took 45 seconds to link.
MSVC link took ~2 sec.
LLVM for Windows (Visual Studio LLVM plugin) ~10 sec in total (link time unkown).

Machine: i7-4771, 16GB RAM, SSD
